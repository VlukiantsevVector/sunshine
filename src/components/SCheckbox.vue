<template>
  <label
    :class="classes(modifiers)"
    @mousedown.native="$event.preventDefault()"
  >
    <input
      type="checkbox"
      :disabled="inactive"
      v-model="isChecked"
      @focus="hasFocus = true"
      @blur="hasFocus = false"
    >
    <s-checkable
      :inactive="inactive"
      :checked="isChecked"
      :has-focus="hasFocus"
    />
    <span>
      <slot />
    </span>
  </label>
</template>

<script lang="ts">
import Vue from 'vue';
import mixins from 'vue-typed-mixins';
import SCheckable from './SCheckable.vue';
import { ClassesMixin } from '../lib/utils';

export default mixins(ClassesMixin).extend({
  name: 's-checkbox',

  components: {
    SCheckable,
  },

  model: {
    prop: 'checked',
    event: 'change',
  },

  props: {
    checked: {
      type: Boolean,
      required: false,
    },

    inactive: {
      type: Boolean,
      default: false,
    },
  },

  data() {
    return {
      hasFocus: false,
      isChecked: this.checked === null ? null : !!this.checked,
    };
  },

  watch: {
    checked(val) {
      this.isChecked = val === null ? null : !!val;
    },

    isChecked(val) {
      if (val !== this.checked) {
        this.$emit('change', val);
      }
    },
  },

  computed: {
    modifiers(): object {
      return {
        focus: this.hasFocus,
        inactive: this.inactive,
      };
    },
  },
});
</script>
