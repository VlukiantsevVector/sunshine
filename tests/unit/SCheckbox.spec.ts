import { mount, config } from '@vue/test-utils';
import SCheckbox from '@/components/SCheckbox.vue';

config.logModifiedComponents = false;

describe('SCheckbox', () => {
  it('renders checkbox', () => {
    const wrapper = mount(SCheckbox, {
      propsData: {
        checked: true,
      },
      slots: {
        default: 'I am a checkbox',
      },
    });
    expect(wrapper.contains('.s-checkable__checkmark')).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('changes state', () => {
    const wrapper = mount(SCheckbox, {
      slots: {
        default: 'I am a checkbox',
      },
    });

    expect(wrapper.contains('.s-checkable__checkmark')).toBeFalsy();
    wrapper.element.click();
    expect(wrapper.contains('.s-checkable__checkmark')).toBeTruthy();
  });

  it('shows multiple state', () => {
    const wrapper = mount(SCheckbox, {
      propsData: {
        checked: null,
      },
      slots: {
        default: 'I am a checkbox',
      },
    });

    expect(wrapper.contains('.s-checkable__checkmark')).toBeTruthy();
  });

  it('emits input', () => {
    const wrapper = mount(SCheckbox, {
      propsData: {
        checked: false,
      },
      slots: {
        default: 'I am a checkbox',
      },
    });

    wrapper.trigger('click');
    expect(wrapper.emitted('change')[0]).toEqual([true]);
  });
});
