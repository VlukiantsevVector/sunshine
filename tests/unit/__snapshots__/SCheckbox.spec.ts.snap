// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SCheckbox renders checkbox 1`] = `<label class="s-checkbox"><input type="checkbox"> <span class="s-checkable"><svg version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" class="s-svg s-checkable__checkmark" name="fade"><g></g></svg></span> <span>I am a checkbox</span></label>`;
